### PR TITLE
Add threadfence to 2-stage reduction for correct writes visibility

### PR DESCRIPTION
**Pull request by ngimel**
Final block accumulating 2-stage reduction result has to complete acquire pattern to make sure the writes of all other blocks are visible to it, see https://docs.nvidia.com/cuda/parallel-thread-execution/index.html?highlight=atom#release-and-acquire-patterns

### CODE REVIEW

**Comment by eqy** [comment]
Was there a failing repro for this that should also be checked in?

**Comment by ngimel** [comment]
Repro requires at least 8 gpus and running for a few minutes, and also has triton dependency, so can't really be turned into a test

**Review by eqy — APPROVED** [review]

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/native/cuda/Reduce.cuh [hunk 1]
```diff
@@ -807,6 +807,7 @@ struct ReduceOp {
     bool is_last_block_done = mark_block_finished();
 
     if (is_last_block_done) {
+      __threadfence(); // complete the aquire pattern after atomic
```
Comment (christopherhesse) [inline aten/src/ATen/native/cuda/Reduce.cuh:810]: There's a typo in `acquire`

Comment (ngimel) [inline aten/src/ATen/native/cuda/Reduce.cuh:810]:
```suggestion
      __threadfence(); // complete the acquire pattern after atomic
```

**Pull request closed without merging**